API/hermes_abi/hermes_abi.h

```diff
@@ -0,0 +1,400 @@
+
+enum HermesABIValueKind {
+  HermesABIValueKindUndefined = 0,
+  HermesABIValueKindNull = 1,
+  HermesABIValueKindBoolean = 2,
+  HermesABIValueKindError = 3,
+  HermesABIValueKindNumber = 4,
+  HermesABIValueKindSymbol = 5 | HERMES_ABI_POINTER_MASK,
+  HermesABIValueKindBigInt = 6 | HERMES_ABI_POINTER_MASK,
+  HermesABIValueKindString = 7 | HERMES_ABI_POINTER_MASK,
+  HermesABIValueKindObject = 9 | HERMES_ABI_POINTER_MASK,
+};
+
+struct HermesABIValue {
+  HermesABIValueKind kind;
+  union {
+    bool boolean;
+    double number;
+    HermesABIManagedPointer *pointer;
+    HermesABIErrorCode error;
+  } data;
+};
+struct HermesABIValueOrError {
+  HermesABIValue value;
+};
+
+struct HermesABIASCIIRef {
+  const char *data;
+  size_t length;
+};
+
+/// Define the structure for buffers containing JS source or bytecode. This is
+/// designed to mirror the functionality of jsi::Buffer.
+struct HermesABIBufferVTable {
+  void (*release)(HermesABIBuffer *);
+};
+struct HermesABIBuffer {
+  const HermesABIBufferVTable *vtable;
+  const uint8_t *data;
+  size_t size;
+};
+
+struct HermesABIByteBufferVTable {
+  void (*grow_by)(HermesABIByteBuffer *, size_t);
+};
+struct HermesABIByteBuffer {
+  const HermesABIByteBufferVTable *vtable;
+  uint8_t *data;
+  size_t available;
+};
+
+/// Define the structure for host functions. This is designed to recreate the
+/// functionality of jsi::HostFunction.
+struct HermesABIHostFunctionVTable {
+  HermesABIValueOrError (*call)(
+      HermesABIHostFunction *,
+      HermesABIContext *,
+      const HermesABIValue *,
+      const HermesABIValue *,
+      size_t);
+  void (*release)(HermesABIHostFunction *);
+};
+struct HermesABIHostFunction {
+  const HermesABIHostFunctionVTable *vtable;
+};
+
+/// Define the structure for lists of PropNameIDs, so that they can be returned
+/// by get_property_names on a HostObject.
+struct HermesABIPropNameIDListVTable {
+  void (*release)(HermesABIPropNameIDList *);
+};
+struct HermesABIPropNameIDList {
+  const HermesABIPropNameIDListVTable *vtable;
+  const HermesABIPropNameID *props;
+  size_t size;
+};
+
+/// Define the structure for host objects. This is designed to recreate the
+/// functionality of jsi::HostObject.
+struct HermesABIHostObjectVTable {
+  HermesABIValueOrError (
+      *get)(HermesABIHostObject *, HermesABIContext *, HermesABIPropNameID);
+  HermesABIVoidOrError (*set)(
+      HermesABIHostObject *,
+      HermesABIContext *,
+      HermesABIPropNameID,
+      const HermesABIValue *);
+  HermesABIPropNameIDListPtrOrError (
+      *get_property_names)(HermesABIHostObject *, HermesABIContext *);
+  void (*release)(HermesABIHostObject *);
+};
+struct HermesABIHostObject {
+  const HermesABIHostObjectVTable *vtable;
+};
+
+struct HermesABINativeStateVTable {
+  void (*release)(HermesABINativeState *);
+};
+struct HermesABINativeState {
+  const HermesABINativeStateVTable *vtable;
+};
+
+struct HermesABIVTable {
```

Comment:
Should we name it HermesABIContextVTable? To match the naming pattern for the code above.

Reply:
I view this as the list of exported functions for our ABI, rather than as the vtable for `HermesABIContext`. `HermesABIContext` doesn't actually have a vtable, since it is just an opaque pointer.

But that raises the reasonable question of why `HermesABIContext` is meaningfully different than the other types which do define vtables. For most of those types, they are vtables that are fed into the ABI from outside (e.g. `HostObject`), so it is reasonable that integrators may want to create different instances with different vtables. This property is useful in practice to check whether a `HostObject` was in fact created by the current C++ wrapper (see the `isHostObject` implementation).

The only exception is `HermesABIManagedPointer`'s `invalidate` function. For that, it is reasonable that the implementation may provide different wrappers for different types (e.g. `Object` vs `WeakObject`). In addition, given that `jsi::Runtime::PointerValue` can be invalidated using just a pointer to it, if we want to replace it, we have to recreate that behaviour.